crates/bevy_math/src/curve/mod.rs

```diff
@@ -0,0 +1,610 @@
+//! The [`Curve`] trait, used to describe curves in a number of different domains. This module also
+//! contains the [`Interval`] type, along with a selection of core data structures used to back
+//! curves that are interpolated from samples.
+
+pub mod interval;
+
+pub use interval::{everywhere, interval, Interval};
+
+use interval::{InfiniteIntervalError, InvalidIntervalError};
+use std::{marker::PhantomData, ops::Deref};
+use thiserror::Error;
+
+#[cfg(feature = "bevy_reflect")]
+use bevy_reflect::Reflect;
+
+/// A trait for a type that can represent values of type `T` parametrized over a fixed interval.
+/// Typical examples of this are actual geometric curves where `T: VectorSpace`, but other kinds
+/// of interpolable data can be represented instead (or in addition).
+pub trait Curve<T> {
+    /// The interval over which this curve is parametrized.
+    fn domain(&self) -> Interval;
+
+    /// Sample a point on this curve at the parameter value `t`, extracting the associated value.
+    fn sample(&self, t: f32) -> T;
+
+    /// Sample a point on this curve at the parameter value `t`, returning `None` if the point is
+    /// outside of the curve's domain.
+    fn sample_checked(&self, t: f32) -> Option<T> {
+        match self.domain().contains(t) {
+            true => Some(self.sample(t)),
+            false => None,
+        }
+    }
+
+    /// Sample a point on this curve at the parameter value `t`, clamping `t` to lie inside the
+    /// domain of the curve.
+    fn sample_clamped(&self, t: f32) -> T {
+        let t = self.domain().clamp(t);
+        self.sample(t)
+    }
+
+    /// Create a new curve by mapping the values of this curve via a function `f`; i.e., if the
+    /// sample at time `t` for this curve is `x`, the value at time `t` on the new curve will be
+    /// `f(x)`.
+    fn map<S, F>(self, f: F) -> MapCurve<T, S, Self, F>
+    where
+        Self: Sized,
+        F: Fn(T) -> S,
```

Comment:
why take a function and not another curve..

edit: i see there's a `reparameterize_by_curve` below, but no `map_by_curve`.

i think not implementing curve for `T: Fn(f32) -> f32` makes sense, but i would like to suggest a rework of this interface:

1. change `map` and `reparameterise` to take curves instead of functions.

2. introduce `map_with` and `reparameterize_with` methods which take functions.

edit 2: at i see it now! curves have to take `f32` as an input, but don't have to return it, so what i suggested is quite brittle. ignore me!